camel/models/gemini_model.py

```diff
@@ -0,0 +1,185 @@
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+from typing import Any, Dict, Optional
+
+from camel.configs import Gemini_API_PARAMS
+from camel.models import BaseModelBackend
+from camel.types import (
+    ChatCompletion,
+    ChatCompletionMessage,
+    Choice,
+    ModelType,
+)
+from camel.utils import (
+    BaseTokenCounter,
+    GeminiTokenCounter,
+    model_api_key_required,
+)
+
+
+class GeminiModel(BaseModelBackend):
+    r"""Gemini API in a unified BaseModelBackend interface."""
+
+    def __init__(
+        self,
+        model_type: ModelType,
+        model_config_dict: Dict[str, Any],
+        api_key: Optional[str] = None,
+        url: Optional[str] = None,
+    ) -> None:
+        r"""Constructor for Gemini backend.
+
+        Args:
+            model_type (ModelType): Model for which a backend is created
+            model_config_dict (Dict[str, Any]): A dictionary that will
+                be fed into generate_content().
+            api_key (Optional[str]): The API key for authenticating with the
+                gemini service. (default: :obj:`None`)
+        """
+        import os
+
+        import google.generativeai as genai
+        from google.generativeai.types.generation_types import GenerationConfig
+
+        super().__init__(model_type, model_config_dict, api_key, url)
+        self._api_key = api_key or os.environ.get("GOOGLE_API_KEY")
+        genai.configure(api_key=self._api_key)
+        self._client = genai.GenerativeModel(model_type.value)
+        self._token_counter: Optional[BaseTokenCounter] = None
+        keys = list(self.model_config_dict.keys())
+        generation_config_dict = {
+            k: self.model_config_dict.pop(k)
+            for k in keys
+            if hasattr(GenerationConfig, k)
+        }
+        generation_config = genai.types.GenerationConfig(
+            **generation_config_dict
+        )
+        self.model_config_dict["generation_config"] = generation_config
+
+    @property
+    def token_counter(self) -> BaseTokenCounter:
+        if not self._token_counter:
+            self._token_counter = GeminiTokenCounter(self.model_type)
+        return self._token_counter
+
+    @model_api_key_required
+    def run(self, contents):
+        r"""Runs inference of Gemini model.
+        This method can handle multimodal input
+
+        Args:
+            contents: Message list or Message with the chat history
+            in Gemini API format or OpenAi format.
+            example: contents = [{'role':'user', 'parts': ['hello']}]
+
+
+        Returns:
+            response: A ChatCompletion object formatted for the OpenAI API.
+
+        If it is not in streaming mode,
+        you can directly output the response.text.
+
+        If it is in streaming mode,
+        you can iterate over the response chunks as they become available.
+        """
```

Comment:
if it's in stream mode, we may also need to convert the response into OpenAI format